docs/proposals/202105_add_function_framework.md

```diff
@@ -0,0 +1,435 @@
+    userFunction "{{.Package.Name}}" 
+    functionFramework "example.com/hello/pkg/offf-go"
+    "fmt"
+    cloudevents "github.com/cloudevents/sdk-go/v2"
+    "log"
+    "net/http"
+    "os"
+)
+
+func register(fn interface{}) error {
+    ctx := context.Background()
+    if fnHTTP, ok := fn.(func (http.ResponseWriter, *http.Request)); ok {
+        if err := functionFramework.RegisterHTTPFunction(ctx, "/", fnHTTP); err != nil {
+            return fmt.Errorf("Function failed to register: %v\n", err)
+        }
+    } else if fnCloudEvent, ok := fn.(func (context.Context, cloudevents.Event)); ok {
+        if err := functionFramework.RegisterCloudEventFunction(ctx, "/", fnCloudEvent); err != nil {
+            return fmt.Errorf("Function failed to register: %v\n", err)
+        }
+    }
+    return nil
+}
+
+func main() {
+    if err := register(userFunction.{{.Package.Target}}); err != nil {
+        log.Fatalf("Failed to register: %v\n", err)
+    }
+
+    // Use PORT environment variable, or default to 8080.
+    port := "8080"
+    if envPort := os.Getenv("PORT"); envPort != "" {
+       port = envPort
+    }
+    if err := functionFramework.Start(port); err != nil {
+       log.Fatalf("Failed to start: %v\n", err)
+    }
+}
+```
+
+#### Function Context
+
+Use the context structure to relate the definition of function crd and further provide the relevant tool suite to enable user to develop the code in a friendly, abstract form.
+
+Here is an example. The user wants its function to fetch data from the input source, process it, and then send the result to the output target. Note that there may be multiple outputs.
+
+```go
+func BindingsFunction(w http.ResponseWriter, r *http.Request) {
+    content, err := io.ReadAll(r.Body)
+
+    type Data struct {
+        OrderId int `json:"order_id"`
+        Content string `json:"content"`
+    }
+    type payload struct {
+        Data *Data `json:"data"`
+        Operation string `json:"operation"`
+    }
+
+    n := 0
+
+    for {
+        n++
+        p := &payload{}
+        p.Data = &Data{
+            OrderId: n,
+            Content: content.(string),
+        }
+        p.Operation = "create"
+        time.Sleep(1 * time.Second)
+
+        _, err = http.Post("http://output.target", "application/json", strings.NewReader(string(body)))
+        if err != nil {
+            fmt.Errorf("Error: %v\n", err)
+        }
+    }
+
+    if err != nil {
+        w.WriteHeader(500)
+    } else {
+        w.WriteHeader(200)
+    }
+}
+```
+
+The role of the Function Context is to pass information about the resources reconcile by the function crd. For example, when you use Dapr, the context can associate some properties of the bindings component to your code.
+
+The following is an example of the structure of a Function Context.
+
+```go
+type Output struct {
+    Url string `json:"url"`
+    Method HttpMethod `json:"method"`
+}
+
+type Input struct {
+    Enable InputOutputSwitch `json:"enable"`
```

Comment:
Enable => Enabled
simply a bool type is enough, pls refer to https://github.com/kubesphere/notification-manager/blob/master/pkg/apis/v2beta2/receiver_types.go#L42

Reply:
ok